src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,141 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private transient Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int initCapacity) {
+        if (initCapacity <= 0) {
+            throw new IllegalArgumentException(
+                    "Can't create array with negative length");
+        }
+        elements = new Object[initCapacity];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfNeeded();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        addElementArrayListIndexOutOfBoundException(index);
+        growIfNeeded();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfNeeded(list);
+        int listSize = list.size();
+        for (int i = 0; i < listSize; i++) {
+            elements[size + i] = list.get(i);
+        }
+        size += listSize;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        arrayListIndexOutOfBoundException(index);
+        return (T)elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        arrayListIndexOutOfBoundException(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        arrayListIndexOutOfBoundException(index);
+        T removedElement = (T)elements[index];
+        removeElement(index);
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int indexRemovedElement = -1;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null
+                    && elements[i].equals(element)
+                    || elements[i] == element) {
+                indexRemovedElement = i;
+                break;
+            }
+        }
+        if (indexRemovedElement == -1) {
+            throw new NoSuchElementException("Can't find element: " + element);
+        }
+        T removedElement = (T)elements[indexRemovedElement];
+        removeElement(indexRemovedElement);
+        return removedElement;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void growIfNeeded() {
+        if (elements.length == 1 && elements.length == size) {
```

Comment:
it is possible to insert this if condiotion after  if (elements.length == size), so It will be nested if first one will be if (elements.length == size) and than if (elements.length == 1)

Reply:
Agree with Andrii. Do sth like 
```
if (first condition) {
  if (second condition) {
    ...
  } else {
    ...
  }
}
```
instead of do if-if containing a part of the same conditions